File: src/main/java/com/marsraver/LedFx/layout/LayoutLoader.java
package com.marsraver.LedFx.layout;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * Loads layout configurations from XML files.
 * Parses XML files in the resources/layouts directory.
 */
public class LayoutLoader {
    
    private static final String LAYOUTS_PATH = "/layouts/";
    
    /**
     * Loads a layout configuration from an XML file.
     * 
     * @param layoutName the name of the layout file (without .xml extension)
     * @return the parsed layout configuration
     * @throws Exception if the layout file cannot be loaded or parsed
     */
    public static LayoutConfig loadLayout(String layoutName) throws Exception {
        String resourcePath = LAYOUTS_PATH + layoutName + ".xml";
        InputStream inputStream = LayoutLoader.class.getResourceAsStream(resourcePath);
        
        if (inputStream == null) {
            throw new IllegalArgumentException("Layout file not found: " + resourcePath);
        }
        
        try {
            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
            DocumentBuilder builder = factory.newDocumentBuilder();
            Document document = builder.parse(inputStream);
            
            return parseLayout(document);
        } finally {
            inputStream.close();
        }
    }
    
    /**
     * Parses a layout configuration from a DOM document.
     */
    private static LayoutConfig parseLayout(Document document) {
        Element root = document.getDocumentElement();
        
        if (!"layout".equals(root.getTagName())) {
            throw new IllegalArgumentException("Root element must be 'layout'");
        }
        
        LayoutConfig layout = new LayoutConfig();
        
        // Parse layout attributes
        layout.setName(getAttributeValue(root, "name", "Unnamed"));
        layout.setTitle(getAttributeValue(root, "title", "LedFx"));
        layout.setWindowWidth(getIntAttribute(root, "windowWidth", 500));
        layout.setWindowHeight(getIntAttribute(root, "windowHeight", 400));
        
        // Parse grids
        NodeList gridNodes = root.getElementsByTagName("grid");
        for (int i = 0; i < gridNodes.getLength(); i++) {
            Element gridElement = (Element) gridNodes.item(i);
            GridConfig grid = parseGrid(gridElement);
            layout.addGrid(grid);
        }
        
        return layout;
    }
    
    /**
     * Parses a single grid configuration from an XML element.
     */
    private static GridConfig parseGrid(Element gridElement) {
        GridConfig grid = new GridConfig();
        
        grid.setId(getAttributeValue(gridElement, "id", "grid"));
        grid.setDeviceIp(getAttributeValue(gridElement, "deviceIp", "192.168.1.100"));
        grid.setLedCount(getIntAttribute(gridElement, "ledCount", 256));
        grid.setX(getIntAttribute(gridElement, "x", 0));
        grid.setY(getIntAttribute(gridElement, "y", 0));
        grid.setWidth(getIntAttribute(gridElement, "width", 240));
        grid.setHeight(getIntAttribute(gridElement, "height", 240));
        grid.setGridSize(getIntAttribute(gridElement, "gridSize", 16));
        grid.setPixelSize(getIntAttribute(gridElement, "pixelSize", 15));
        
        return grid;
    }
    
    /**
     * Gets a string attribute value with a default.
     */
    private static String getAttributeValue(Element element, String attributeName, String defaultValue) {
        String value = element.getAttribute(attributeName);
        return value.isEmpty() ? defaultValue : value;
    }
    
    /**
     * Gets an integer attribute value with a default.
     */
    private static int getIntAttribute(Element element, String attributeName, int defaultValue) {
        String value = element.getAttribute(attributeName);
        try {
            return value.isEmpty() ? defaultValue : Integer.parseInt(value);
        } catch (NumberFormatException e) {
            System.err.println("Invalid integer value for attribute '" + attributeName + "': " + value);
            return defaultValue;
        }
    }
    
    /**
     * Lists all available layout files.
     * 
     * @return list of layout names (without .xml extension)
     */
    public static List<String> listAvailableLayouts() {
        List<String> layouts = new ArrayList<>();
        
        // For now, we'll hardcode the known layouts
        // In a more sophisticated implementation, we could scan the resources directory
        layouts.add("TwoGrids");
        
        return layouts;
    }
}

File: src/main/kotlin/com/marsraver/LedFx/layout/LayoutLoader.kt
package com.marsraver.LedFx.layout

import com.marsraver.LedFx.wled.ColorMapping
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import org.w3c.dom.Document
import org.w3c.dom.Element
import java.util.*
import javax.xml.parsers.DocumentBuilderFactory

/**
 * Loads layout configurations from XML files.
 * Parses XML files in the resources/layouts directory.
 */
object LayoutLoader {
    private const val LAYOUTS_PATH = "/layouts/"
    private val log: Logger = LogManager.getLogger(LayoutLoader::class.java)

    /**
     * Loads a layout configuration from an XML file.
     *
     * @param layoutName the name of the layout file (without .xml extension)
     * @return the parsed layout configuration
     * @throws Exception if the layout file cannot be loaded or parsed
     */
    @Throws(Exception::class)
    fun loadLayout(layoutName: String): LayoutConfig {
        val resourcePath = "$LAYOUTS_PATH$layoutName.xml"
        val inputStream = LayoutLoader::class.java.getResourceAsStream(resourcePath)

        requireNotNull(inputStream) { "Layout file not found: $resourcePath" }

        try {
            val factory = DocumentBuilderFactory.newInstance()
            val builder = factory.newDocumentBuilder()
            val document = builder.parse(inputStream)

            return parseLayout(document)
        } finally {
            inputStream.close()
        }
    }

    /**
     * Parses a layout configuration from a DOM document.
     */
    private fun parseLayout(document: Document): LayoutConfig {
        val root = document.documentElement

        require(root.tagName == "layout") { "Root element must be 'layout'" }

        val layout = LayoutConfig()

        // Parse layout attributes
        layout.name = getAttributeValue(root, "name", "Unnamed")
        layout.title = getAttributeValue(root, "title", "LedFx")
        layout.windowWidth = getIntAttribute(root, "windowWidth", 500)
        layout.windowHeight = getIntAttribute(root, "windowHeight", 400)

        // Parse grids
        val gridNodes = root.getElementsByTagName("grid")
        for (i in 0 until gridNodes.length) {
            val gridElement = gridNodes.item(i) as Element
            val grid = parseGrid(gridElement)
            layout.addGrid(grid)
        }

        return layout
    }

    /**
     * Parses a single grid configuration from an XML element.
     */
    private fun parseGrid(gridElement: Element): GridConfig {
        val grid = GridConfig()

        grid.id = getAttributeValue(gridElement, "id", "grid")
        grid.deviceIp = getAttributeValue(gridElement, "deviceIp", "192.168.1.100")
        grid.ledCount = getIntAttribute(gridElement, "ledCount", 256)
        grid.x = getIntAttribute(gridElement, "x", 0)
        grid.y = getIntAttribute(gridElement, "y", 0)
        grid.width = getIntAttribute(gridElement, "width", 240)
        grid.height = getIntAttribute(gridElement, "height", 240)
        grid.gridSize = getIntAttribute(gridElement, "gridSize", 16)
        grid.pixelSize = getIntAttribute(gridElement, "pixelSize", 15)

        // Parse colorMapping attribute
        val colorMappingStr = getAttributeValue(gridElement, "colorMapping", "")
        if (colorMappingStr.isNotEmpty()) {
            try {
                val colorMapping = ColorMapping.valueOf(colorMappingStr.uppercase(Locale.getDefault()))
                grid.colorMapping = colorMapping
            } catch (e: IllegalArgumentException) {
                log.error(
                    "Invalid colorMapping '$colorMappingStr' for grid '${grid.id}', using default GBR"
                )
            }
        }

        return grid
    }

    /**
     * Gets a string attribute value with a default.
     */
    private fun getAttributeValue(element: Element, attributeName: String, defaultValue: String): String {
        val value = element.getAttribute(attributeName)
        return if (value.isEmpty()) defaultValue else value
    }

    /**
     * Gets an integer attribute value with a default.
     */
    private fun getIntAttribute(element: Element, attributeName: String, defaultValue: Int): Int {
        val value = element.getAttribute(attributeName)
        return try {
            if (value.isEmpty()) defaultValue else value.toInt()
        } catch (e: NumberFormatException) {
            log.error("Invalid integer value for attribute '$attributeName': $value")
            defaultValue
        }
    }

    /**
     * Lists all available layout files.
     *
     * @return list of layout names (without .xml extension)
     */
    fun listAvailableLayouts(): MutableList<String> {
        val layouts: MutableList<String> = ArrayList()

        // For now, we'll hardcode the known layouts
        // In a more sophisticated implementation, we could scan the resources directory
        layouts.add("OneGrid")
        layouts.add("TwoGrids")
        layouts.add("FourGrids")

        return layouts
    }
}

